src/core/probe/probe.rs

```diff
@@ -43,14 +44,49 @@ impl Probe {
             kernel::Symbol::Func(_) => bail!("Symbol cannot be probed with a raw tracepoint"),
         }
     }
-}
 
-// Use mem::variant_count::<Probe>() when available in stable.
-pub(crate) const PROBE_VARIANTS: usize = 4;
+    /// Set a probe option.
+    pub(crate) fn set_option(&mut self, opt: &ProbeOption) -> Result<()> {
+        match self {
+            Probe::Kprobe(kp) | Probe::Kretprobe(kp) | Probe::RawTracepoint(kp) => {
+                kp.set_option(opt)
+            }
+            Probe::Usdt(_) => bail!("set_option isn't supported for USDT yet"),
+        }
+    }
+
+    /// Merge two Probe options. We can't merge two different probe variants,
+    /// except for the special kprobe/kretprobe case.
+    pub(crate) fn merge_options(&mut self, other: &Probe) -> Result<()> {
+        match (self, other) {
+            (Probe::Kprobe(kp), Probe::Kprobe(other))
+            | (Probe::Kretprobe(kp), Probe::Kretprobe(other))
+            | (Probe::Kprobe(kp), Probe::Kretprobe(other))
+            | (Probe::Kretprobe(kp), Probe::Kprobe(other))
+            | (Probe::RawTracepoint(kp), Probe::RawTracepoint(other)) => {
+                kp.merge_options(other);
+            }
+            (Probe::Usdt(_), Probe::Usdt(_)) => (),
+            _ => bail!("Can't merge options on different probe variants"),
+        }
+        Ok(())
+    }
+}
 
 impl Probe {
+    /// Use the underlying symbol to get its name and use it as a key which can
+    /// be used to differenciate between probes.
+    pub(crate) fn key(&self) -> String {
+        match &self {
+            Probe::Kprobe(probe) | Probe::Kretprobe(probe) | Probe::RawTracepoint(probe) => {
+                probe.symbol.name()
```

Comment:
I'm afraid this will not work (in fact, it breaks ovs module).
kretprobes and kprobes might have the same name (because they attach to the same symbol), so this will store them under the same key replacing the first probe info with the last one that is added.
I'd suggest making Probe drive from Hash, Eq and PartialEq so it can directly be used in HashMaps